AArch64: Optimize memmove for non-power-of-two sizes

osamakader opened the pull request:
Fixes #165948

osamakader committed:
AArch64: Optimize memmove for non-power-of-two sizes using overlapping loads/stores

This change improves memmove code generation for non-power-of-two sizes
on AArch64 by using overlapping loads/stores instead of mixed-size
operations, matching GCC's approach.

For example, for a 7-byte memmove:
- Before: ldrb + ldrh + ldr (3 loads, 3 stores)
- After:  ldur w8, [x1, #3] + ldr w9, [x1] (2 loads, 2 stores)

The optimization handles sizes 5-65 bytes that are not multiples of 8:
- 5-7 bytes:   two overlapping i32 operations
- 9-15 bytes:  two overlapping i64 operations
- 17-23 bytes: two i64 + one overlapping i64
- 25-31 bytes: one v16i8 vector + one overlapping i64
- 33-47 bytes: two v16i8 vectors + one overlapping i64
- 49-63 bytes: three v16i8 vectors + one overlapping i64
- 65 bytes:    four v16i8 vectors + one overlapping i64

This addresses issue #165948 where LLVM generated suboptimal code
compared to GCC for non-power-of-two memmove sizes.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

nasherm commented on llvm/lib/Target/AArch64/AArch64SelectionDAGInfo.cpp:
```diff
@@ -252,6 +252,332 @@ SDValue AArch64SelectionDAGInfo::EmitTargetCodeForMemmove(
+        isVolatile ? MachineMemOperand::MOVolatile : MachineMemOperand::MONone;
+
+    // For sizes 5-7 bytes: use two overlapping i32 operations
+    if (SizeVal >= 5 && SizeVal <= 7) {
```
The pattern in these ```if``` blocks seems to be repeating the same core logic. Is there anyway way you can generalize it to a function? Something that takes the ```SizeVal``` as a parameter and generates the DAG node from that?

nasherm reviewed (changes requested):
I like the idea of this patch. But I have some comments

The codegen from this patch doesn't seem that much better for testcases ```move33```, ```move49```, and ```move65``` https://godbolt.org/z/nEYvE5YY7

osamakader committed:
AArch64: Optimize memmove for non-power-of-two sizes refactoring

- Refactored repetitive code
- Simplified size check: Using binary arithmetic (Size & (Size - 1)) != 0
- Updated comments

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader commented on llvm/lib/Target/AArch64/AArch64SelectionDAGInfo.cpp:
```diff
@@ -252,6 +252,332 @@ SDValue AArch64SelectionDAGInfo::EmitTargetCodeForMemmove(
+    }
+
+    // For sizes 33-47 bytes: use 2 x v16i8 (vectors) + overlapping i64
+    if (SizeVal >= 33 && SizeVal <= 47) {
```
Actually, we have optimizations for some sizes, e.g.

Size 35: 3 loads/3 stores (optimized) vs 4 loads/4 stores (unoptimized)
Size 47: 3 loads/3 stores (optimized) vs 5 loads/5 stores (unoptimized)

davemgreen reviewed (commented):
I think I would expect this to to work similarly to memcpy, where we already lower to unaligned loads. Providing we do the loads before the stores that sounds like it should be alright.

osamakader commented:
@nasherm can you please re-review?

osamakader commented:
memmove doesn't work the same way as memcpy because of how the generic SelectionDAG path handles volatile operations.

memcpy (non-volatile): In getMemcpyLoadsAndStores (SelectionDAG.cpp), MemOp::Copy uses the actual isVol parameter, so allowOverlap() == true for non-volatile cases. This allows:
Overlapping loads/stores optimization (from findOptimalMemOpLowering)
Chaining optimization (from chainLoadsAndStoresForMemcpy, fixed in PR 168890)

memmove (always treated as volatile): In getMemmoveLoadsAndStores (SelectionDAG.cpp:8799), MemOp::Copy is hardcoded to IsVolatile=true, so allowOverlap() == false. This prevents:
Overlapping loads/stores optimization in the generic path
The generic path falls back to mixed-size operations (e.g., i32 + i16 + i8 for 7 bytes)
Since memmove is hardcoded to volatile in the generic path, it can't use overlapping loads/stores. 

Our AArch64-specific implementation:
Uses overlapping loads/stores (e.g., two i32 operations for 7 bytes instead of i32 + i16 + i8)
Provides better codegen for non-power-of-two sizes.

nasherm reviewed (approved):
LGTM. Sorry for the delay. I was on holiday. I'd wait a few days before merging to see if anyone else has more comments.

davemgreen commented:
> memmove doesn't work the same way as memcpy because of how the generic SelectionDAG path handles volatile operations.
> ...

My point wasn't the low level technical reasons why the code works the way it does at the moment, but the high level structural question of why it cannot work in a similar same way. This seems to add a fair amount of technical debt to the AArch64 backend, where as there is already generic code that knows how to split up overlapping reads/writes more optimally.

clingfei commented:
> > memmove doesn't work the same way as memcpy because of how the generic SelectionDAG path handles volatile operations.
> > ...
> 
> My point wasn't the low level technical reasons why the code works the way it does at the moment, but the high level structural question of why it cannot work in a similar same way. This seems to add a fair amount of technical debt to the AArch64 backend, where as there is already generic code that knows how to split up overlapping reads/writes more optimally.

I'm also curious why the code only optimizes for specific test cases instead of proposing a more general implementation. I think the optimization on such a limited set of cases is meaningless.

By the way, [PR 168890](https://github.com/llvm/llvm-project/pull/168890) hasn't received any replies for a long time. Could someone take a look and offer some suggestions for the next steps?
